pkg/jobs/job_info_storage.go

```diff
@@ -114,12 +107,33 @@ func (i ProgressStorage) Set(
 		ts = resolved.AsOfSystemTime()
 	}
 
-	if _, err := txn.ExecEx(
-		ctx, "write-job-progress-insert", txn.KV(), sessiondata.NodeUserSessionDataOverride,
-		`INSERT INTO system.job_progress (job_id, written, fraction, resolved) VALUES ($1, now(), $2, $3)`,
+	// Optimistically update the job's progress row, but if it does not exist,
+	// insert it. This could have been implemented as a DELETE followed by an
+	// INSERT, but that would have always required two separate SQL operations.
+	rowsUpdated, err := txn.ExecEx(
+		ctx, "write-job-progress-update", txn.KV(), sessiondata.NodeUserSessionDataOverride,
+		`UPDATE system.job_progress SET (written, fraction, resolved) = (now(), $2, $3) WHERE job_id = $1`,
 		i, frac, ts,
-	); err != nil {
-		return err
+	)
```

Comment:
Can we use an `UPSERT` here? Since this table doesn't have a secondary index and you are setting all columns, I think it should end up with a single 1PC Put.